### ISSUE TITLE
acme resolver not working with persistence enabled

### ISSUE THREAD

**Issue by marcofranssen** [traefik/traefik-helm-chart#396]
The acme resolver isn't working with persistence enabled due to file permissions.

See below the log.

```bash
time="2021-03-22T11:03:24Z" level=error msg="The ACME resolver \"le\" is skipped from the resolvers list because: unable to get ACME account: permissions 660 for /data/acme.json are too open, please use 600"
```

This is an abstract of my values.yml

```
additionalArguments:
  - "--serverstransport.insecureskipverify=true"
  - "--certificatesresolvers.le.acme.dnschallenge=true"
  - "--certificatesresolvers.le.acme.email=me@tld.com"
  - "--certificatesresolvers.le.acme.caserver=https://acme-staging-v02.api.letsencrypt.org/directory"
  - "--certificatesresolvers.le.acme.storage=/data/acme.json"

persistence:
  enabled: true

serviceAccountAnnotations:
  eks.amazonaws.com/role-arn: <<my_arn_with_a_policy_for_route53>>
```

**Comment by marcofranssen** [comment]
Seems the lego provider is nicely writing with 600 permissions.

https://github.com/traefik/traefik/blob/1b21f0723fd985fe14bd357b5b9691d780d8378a/pkg/provider/acme/local_store.go#L111

**Comment by marcofranssen** [comment]
I now figured that on initial creation the permissions are there as 600.

When deleting the pod, the deployment automatically creates a new one and mounts the now already existing volume with the acme.json file.

Now the file has 660 permissions and breaks…

**Comment by marcofranssen** [comment]
I found a solution by adding a initContainer.

```yaml
persistence:
  enabled: true

# this is required to ensure the acme.json has the required 600 permissions when remounting the volume
deployment:
  initContainers:
    - name: fix-permissions-acme-json
      image: alpine:3.12.4
      command:
        - chmod
        - "600"
        - /data/acme.json
      volumeMounts:
        - name: data
          mountPath: /data
```

Would a PR be accepted that adds this initcontainer by default when persistence is enabled?

**Comment by jakubhajek** [comment]
As a workaround you can try to enable initcontainers section:

```yaml
 initContainers:
    - name: volume-permissions
      image: busybox:1.31.1
      command: ["sh", "-c", "chmod -Rv 600 /data/*"]
      volumeMounts:
        - name: data
          mountPath: /data
```

Each time when a pod will be recycled the correct permissions (600) will be set.

**Comment by marcofranssen** [comment]
@jakubhajek thanks, I found the solution 2 minutes before you answered here :)

**Comment by sjoerd-dijkstra** [comment]
Experiencing the same problem here, posted it just a few minutes before in the wrong repo 👆.

I can confirm the workaround of @jakubhajek works 👍

**Comment by mloiseleur** [comment]
@Haribo112 See https://github.com/traefik/traefik-helm-chart/blob/master/EXAMPLES.md#use-traefik-native-lets-encrypt-integration-without-cert-manager for a detailed working example.

**Comment by Haribo112** [comment]
When I follow those exact instructions I end up with a pod that cannot start, because the initcontainer keeps crashing. The logs for the init-container reveal: "touch: /data/acme.json: Permission denied
chmod: /data/acme.json: No such file or directory"

also, other issues on this repo and elsewhere seem to indicate that one should never try to `touch` or otherwise modify the `acme.json` file because traefik cannot handle the file being empty.

**Comment by huedaya** [comment]
Experiencing the same issue as @Haribo112, the odd this is `/data/acme.json` never exists. Either the busybox unable to execute the command or it was replaced it by other stuff. I also suspect that the persistent volume is in read-only mode.

I've been tried many things, including putting on `/tmp` which technically writeable, but got nothing. Seems something changed because of Kubernetes upgrade.

I'm using Kubernetes: 1.27.2

**Comment by monnierant** [comment]
Hi.

I exeperienced the same issue as @huedaya  and @Haribo112 .

I found where the issue is but i'm not able to fix it right now.

TL;DR : The /data (or /cert-ssl for me) folder got 600 permission and is root:root, but the busy box and traefik are running as non root containers so they don't get permission to access the json files

Here are the step that i have done to investigate

I have run the following commands into the busybox initContainers 

```Bash
$ id root
uid=0(root) gid=0(root) groups=0(root),10(wheel)
$ ls -l / | grep ssl-certs
drwxr-xr-x    3 root     root          4096 Jul  8 14:47 ssl-certs
$ ls -l /ssl-certs
drwx------    2 root     root         16384 Jul  8 14:47 lost+found
$ whoami 
whoami: unknown uid 65532
$ touch /ssl-certs/acme-staging.json
touch: /ssl-certs/acme-staging.json: Permission denied
$ chmod -v 600 /ssl-certs/acme-staging.json
chmod: /ssl-certs/acme-staging.json: No such file or directory
```

We can see that the issue come from the runing user doesn't get the permissions on the /ssl-cert/ folder to performe changes.

I try to force the init container to run as root:

```yaml
deployment:
  initContainers:
    # The "volume-permissions" init container is required if you run into permission issues.
    # Related issue: https://github.com/containous/traefik/issues/6972
    - name: volume-permissions
      image: busybox:1.31.1
      command: ["sh", "-c", "id root;ls -l /; ls -l /ssl-certs; whoami; touch /ssl-certs/acme-staging.json; chmod -v 600 /ssl-certs/acme-staging.json; touch /ssl-certs/acme-production.json; chmod -v 600 /ssl-certs/acme-production.json"]
      securityContext:
        runAsUser: 0
        runAsNonRoot: false
        allowPrivilegeEscalation: false
      volumeMounts:
        - name: ssl-certs
          mountPath: /ssl-certs
```

And now the chmod are well executed on `Init container` :

```bash
mode of '/ssl-certs/acme-staging.json' changed to 0600 (rw-------)  
mode of '/ssl-certs/acme-production.json' changed to 0600 (rw-------)
```


But now the issue is on the traefik container :

```Bash
time="2023-07-09T07:17:47Z" level=error msg="The ACME resolver \"staging\" is skipped from the resolvers list because: unable to get ACME account: open /ssl-certs/acme-staging.json: permission denied"  
time="2023-07-09T07:17:47Z" level=error msg="The ACME resolver \"production\" is skipped from the resolvers list because: unable to get ACME account: open /ssl-certs/acme-production.json: permission denied"
```

**Comment by icodeforyou-dot-net** [comment]
I just decided to do something else, I am not sure if it solved the issue though because I am not yet trying to actually save a certificate in the file. But the initContainer appears to be working now.

1) I decided to use the traefik container instead of busybox under the assumption that it will have the appropriate user with uid 65532. 

2) Then I create the file as root, but change the ownership to uid 65532.

```yaml
  initContainers:
    - name: volume-permissions
      image: traefik:v2.10.4
      command:
        [
          "sh",
          "-c",
          "touch /data/acme.json; chown 65532 /data/acme.json; chmod -v 600 /data/acme.json",
        ]
      securityContext:
        runAsNonRoot: false
        runAsGroup: 0
        runAsUser: 0
      volumeMounts:
        - name: traefik-data
          mountPath: /data
```

The initContainer no longer crashes at least.

I will let people know once I am actually able verify that I can store my certs this way.

**Comment by icodeforyou-dot-net** [comment]
@huedaya huedaya

I am not totally sure because my experience with Traefik is still limited, but at least I am past the error from inside traefik container. I'd say it stores the cert now. I also killed the pod and it still going.

I'd say it is working now.

**Comment by tssgery** [comment]
> I just decided to do something else, I am not sure if it solved the issue though because I am not yet trying to actually save a certificate in the file. But the initContainer appears to be working now.
> 
> 1. I decided to use the traefik container instead of busybox under the assumption that it will have the appropriate user with uid 65532.
> 2. Then I create the file as root, but change the ownership to uid 65532.
> 
> ```yaml
>   initContainers:
>     - name: volume-permissions
>       image: traefik:v2.10.4
>       command:
>         [
>           "sh",
>           "-c",
>           "touch /data/acme.json; chown 65532 /data/acme.json; chmod -v 600 /data/acme.json",
>         ]
>       securityContext:
>         runAsNonRoot: false
>         runAsGroup: 0
>         runAsUser: 0
>       volumeMounts:
>         - name: traefik-data
>           mountPath: /data
> ```
> 
> The initContainer no longer crashes at least.
> 
> I will let people know once I am actually able verify that I can store my certs this way.

This seems to be working for me, thanks for the suggestion

**Comment by icodeforyou-dot-net** [comment]
I might create a pull request in a week or two to update the infos/comments the helm chart itself. The old solution that isn't working for everyone is still to be found there. So everyone will know in the future how to get around this. Just have to do some more tests.

**Comment by rsinha29** [comment]
i am having the same issue , is there any update on this one ?

**Comment by icodeforyou-dot-net** [comment]
@rsinha29 which is "the same issue"? Did you try my workaround? I did not have time to work on the helm chart itself so far.

**Comment by rsinha29** [comment]
@icodeforyou-dot-net  , forgot to mention here i used your workaround and it worked 
thanks again!!

**Comment by danktankk** [comment]
> I just decided to do something else, I am not sure if it solved the issue though because I am not yet trying to actually save a certificate in the file. But the initContainer appears to be working now.
> 
> 1. I decided to use the traefik container instead of busybox under the assumption that it will have the appropriate user with uid 65532.
> 2. Then I create the file as root, but change the ownership to uid 65532.
> 
> ```yaml
>   initContainers:
>     - name: volume-permissions
>       image: traefik:v2.10.4
>       command:
>         [
>           "sh",
>           "-c",
>           "touch /data/acme.json; chown 65532 /data/acme.json; chmod -v 600 /data/acme.json",
>         ]
>       securityContext:
>         runAsNonRoot: false
>         runAsGroup: 0
>         runAsUser: 0
>       volumeMounts:
>         - name: traefik-data
>           mountPath: /data
> ```
> 
> The initContainer no longer crashes at least.
> 
> I will let people know once I am actually able verify that I can store my certs this way.

Is this the answer to the issue currently;  the "workaround?"  It seems like a long time to have this issue, but it definitely still exists and has wasted a lot of my day so far.

**Comment by icodeforyou-dot-net** [comment]
Well it is my answer at least :smile: 

It is working, right? I believe it should technically also work with the `busybox` container because `chown` with userid should change the file permissions regardless of the user exists or not. I just used the `traefik` container as that one should actually have a user with id `65532`.

I don't really know when I have the leisure to open a pull request on this one. I would need to test this with the `busybox` container.

**Comment by carpodiem** [comment]
> Well it is my answer at least 😄
> 
> It is working, right? I believe it should technically also work with the `busybox` container because `chown` with userid should change the file permissions regardless of the user exists or not. I just used the `traefik` container as that one should actually have a user with id `65532`.
> 
> But I am using _should_ here for a reason. I didn't actually check any of that.
> 
> I don't really know when I have the leisure to open a pull request on this one. I would need to test this with the `busybox` container.

I have tried with `busybox` and it worked fine too. Probably you will want to change permissions for the whole directory (not only for acme.json)
```sh
✗ kubectl logs traefik-dsfkj874-hfjei -n kube-system -c volume-permissions
mode of '/data/acme.json' changed to 0600 (rw-------)
```

```yaml
 initContainers:
    - name: volume-permissions
      image: busybox:latest
      command:
        [
          "sh",
          "-c",
          "touch /data/acme.json; chown 65532 /data/acme.json; chmod -v 600 /data/acme.json",
        ]
      securityContext:
        runAsNonRoot: false
        runAsGroup: 0
        runAsUser: 0
      volumeMounts:
        - name: data
          mountPath: /data
```

**Comment by life5ign** [comment]
TLDR proposed workaround doesn't work in chart version 26.0.0 (possibly others too) since setting uid 0 violates the pod security context.

per
> 1. I decided to use the traefik container instead of busybox under the assumption that it will have the appropriate user with uid 65532.
> 2. Then I create the file as root, but change the ownership to uid 65532.
> 
> ```yaml
>   initContainers:
>     - name: volume-permissions
>       image: traefik:v2.10.4
>       command:
>         [
>           "sh",
>           "-c",
>           "touch /data/acme.json; chown 65532 /data/acme.json; chmod -v 600 /data/acme.json",
>         ]
>       securityContext:
>         runAsNonRoot: false
>         runAsGroup: 0
>         runAsUser: 0
>       volumeMounts:
>         - name: traefik-data
>           mountPath: /data
> ```
> 
> The initContainer no longer crashes at least.
> 

This doesn't work for me, since the default chart values (which I haven't changed), for chart version 26.0.0, specify a pod security context, which makes the uid of 0 not allowed for the volume-permissions init container:

The values:
```yaml
podSecurityContext:
  # /!\ When setting fsGroup, Kubernetes will recursively change ownership and
  # permissions for the contents of each volume to match the fsGroup. This can
  # be an issue when storing sensitive content like TLS Certificates /!\
  # fsGroup: 65532
  # -- Specifies the policy for changing ownership and permissions of volume contents to match the fsGroup.
  fsGroupChangePolicy: "OnRootMismatch"
  # -- The ID of the group for all containers in the pod to run as.
  runAsGroup: 65532
  # -- Specifies whether the containers should run as a non-root user.
  runAsNonRoot: true
  # -- The ID of the user for all containers in the pod to run as.
  runAsUser: 65532
```

The error:
` k describe -n traefik pods <your_pod_id>`

>   Normal   Pulled     68s                  kubelet            Successfully pulled image "busybox:latest" in 551ms (551ms including waiting)
  ***Warning  Failed     56s (x8 over 2m19s)  kubelet            Error: container's runAsUser breaks non-root policy (pod: "traefik-blah_traefik(blah)", container: volume-permissions)***
  Normal   Pulled     56s                  kubelet            Successfully pulled image "busybox:latest" in 544ms (544ms including waiting)
  Normal   Pulling    41s (x9 over 2m20s)  kubelet            Pulling image "busybox:latest"

**Comment by life5ign** [comment]
Per my last comment, I just figured out the solution: `fsGroup` was commented out; we need it to be applied so that k8s changes the permissions on volumes to the id specified, before we try to us an initContainer with that id (don't need to run as root with id 0 anymore, if we do this).  So the final config that worked was:

# uncomment fsGroup
```yaml
podSecurityContext:
  # /!\ When setting fsGroup, Kubernetes will recursively change ownership and
  # permissions for the contents of each volume to match the fsGroup. This can
  # be an issue when storing sensitive content like TLS Certificates /!\
  fsGroup: 65532
  # -- Specifies the policy for changing ownership and permissions of volume contents to match the fsGroup.
  fsGroupChangePolicy: "OnRootMismatch"
  # -- The ID of the group for all containers in the pod to run as.
  runAsGroup: 65532
  # -- Specifies whether the containers should run as a non-root user.
  runAsNonRoot: true
  # -- The ID of the user for all containers in the pod to run as.
  runAsUser: 65532
```

# don't use id 0 anymore; just use 65532
```yaml
initContainers:
  # The "volume-permissions" init container is required if you run into permission issues.
  # Related issue: https://github.com/traefik/traefik-helm-chart/issues/396
  - name: volume-permissions
    image: busybox:latest
    command: ["sh", "-c", "touch /data/acme.json; chmod -v 600 /data/acme.json"]
    securityContext:
      runAsNonRoot: true
      runAsGroup: 65532
      runAsUser: 65532
    volumeMounts:
      - name: data
        mountPath: /data
```
Then, we can see success:
```bash
 k logs -n traefik pods/<your_pod> -c volume-permissions                                                                                                   
```
outputs
>mode of '/data/acme.json' changed to 0600 (rw-------)

**Comment by igor9silva** [comment]
I did not manage to make @life5ign solution work for me (26.0.0), but got it working.

For reference, I'm doing this for EKS Fargate:

`traefik-values.yaml`
```yaml
persistence:
  # -- Enable persistence using Persistent Volume Claims
  # ref: http://kubernetes.io/docs/user-guide/persistent-volumes/
  # It can be used to store TLS certificates, see `storage` in certResolvers
  enabled: true
  size: 128Mi
  storageClass: efs
  accessMode: ReadWriteOnce
...
deployment:
   ...
  initContainers:
    # The "volume-permissions" init container is required if you run into permission issues.
    # Related issue: https://github.com/traefik/traefik-helm-chart/issues/396
    - name: volume-permissions
      image: busybox:1.36
      command:
        ["sh", "-c", "touch /data/acme.json; chmod -v 600 /data/acme.json"]
      securityContext:
        runAsNonRoot: false
        runAsGroup: 0
        runAsUser: 0
      volumeMounts:
        - name: data
          mountPath: /data
...
podSecurityContext:
  # /!\ When setting fsGroup, Kubernetes will recursively change ownership and
  # permissions for the contents of each volume to match the fsGroup. This can
  # be an issue when storing sensitive content like TLS Certificates /!\
  # fsGroup: 65532
  # -- Specifies the policy for changing ownership and permissions of volume contents to match the fsGroup.
  fsGroupChangePolicy: "OnRootMismatch"
  # -- The ID of the group for all containers in the pod to run as.
  runAsGroup: 65532
  # -- Specifies whether the containers should run as a non-root user.
  runAsNonRoot: true
  # -- The ID of the user for all containers in the pod to run as.
  runAsUser: 65532
```


`file-system.yaml`
```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: efs
provisioner: efs.csi.aws.com
---
# Traefik will use this to store TLS certificates
apiVersion: v1
kind: PersistentVolume
metadata:
  name: traefik-efs
spec:
  capacity:
    storage: 1Gi
  volumeMode: Filesystem
  accessModes:
    - ReadWriteOnce # Many is only available to dynamic volumes
  persistentVolumeReclaimPolicy: Retain # so disk is not delete after pod deletion
  storageClassName: efs
  csi:
    driver: efs.csi.aws.com
    # As of today, Amazon EFS CSI driver doesn't support dynamic provisioning on Fargate,
    # so we're manually deploying an EFS instance and hardcoding it here 😭
    #
    # Ref: https://github.com/kubernetes-sigs/aws-efs-csi-driver
    volumeHandle: fs-0cbf8b13198606563
```

**Comment by DrummyFloyd** [comment]
This issue seems to be still present .. in the actual Helm chart 26.1.0

**Comment by bermeom** [comment]
By simply changing the ownership of the acme.json file to ID user of traefik pod, the problem was resolved, i.e.,
```yaml
deployment:
   ...
  initContainers:
    # The "volume-permissions" init container is required if you run into permission issues.
    # Related issue: https://github.com/traefik/traefik-helm-chart/issues/396
    - name: volume-permissions
      image: busybox:1.36
      command:       
       ["sh", "-c", "touch /data/acme.json; chmod -v 600 /data/acme.json; adduser -S 65532 65532; chown -R 65532:65532 /data/acme.json"]
```

**Comment by rptaylor** [comment]
Looks like this is the same as https://github.com/traefik/traefik-helm-chart/issues/164

**Comment by ssbagwe** [comment]
InitContainer is not required. Helm chart v37.1.0

**Solution**:

```
  podSecurityContext:
    runAsGroup: 65532
    runAsNonRoot: true
    runAsUser: 65532
    fsGroup: 65532
    fsGroupChangePolicy: "OnRootMismatch"
```
